gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+            rconfig.set_offsets(&[gpio as u32]);
+
+            state.request = Some(Arc::new(
+                self.chip
+                    .request_lines(&rconfig, &config)
+                    .map_err(Error::GpiodFailed)?,
+            ));
+        }
+
+        Ok(())
+    }
+
+    fn get_value(&self, gpio: u16) -> Result<u8> {
+        let state = &self.state[gpio as usize].read().unwrap();
+
+        if let Some(request) = &state.request {
+            Ok(request.get_value(gpio as u32).map_err(Error::GpiodFailed)? as u8)
+        } else {
+            Err(Error::GpioDirectionInvalid(
+                VIRTIO_GPIO_DIRECTION_NONE as u32,
+            ))
+        }
+    }
+
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()> {
+        let state = &self.state[gpio as usize].read().unwrap();
+
+        // Direction change can follow value change, don't fail here for invalid
+        // direction.
+        if let Some(request) = &state.request {
+            request
+                .set_value(gpio as u32, value as i32)
+                .map_err(Error::GpiodFailed)?;
+        }
+
+        Ok(())
+    }
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()> {
+        let state = &mut self.state[gpio as usize].write().unwrap();
+        let mut config = LineConfig::new().map_err(Error::GpiodFailed)?;
+
+        match value as u16 {
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING => config.set_edge_detection(Edge::Rising),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING => config.set_edge_detection(Edge::Falling),
+            VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH => config.set_edge_detection(Edge::Both),
+
+            // Drop the buffer.
+            VIRTIO_GPIO_IRQ_TYPE_NONE => {
+                state.buffer = None;
+                return Ok(());
+            }
+
+            // Only edge IRQs are supported for now.
+            _ => return Err(Error::GpioIrqTypeNotSupported(value)),
+        };
+
+        // Allocate the buffer and configure the line for interrupt.
+        if state.request.is_none() {
+            Err(Error::GpioLineRequestNotConfigured)
+        } else {
+            state.buffer = Some(EdgeEventBuffer::new(1).map_err(Error::GpiodFailed)?);
+
+            state
+                .request
+                .as_ref()
+                .unwrap()
+                .reconfigure_lines(&config)
+                .map_err(Error::GpiodFailed)
+        }
+    }
+
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()> {
+        // While waiting here for the interrupt to occur, it is possible that we receive another
+        // request to disable the interrupt instead, via a call to set_irq_type().
+        //
+        // That request shouldn't wait for this wait_for_interrupt() function to finish, else we
+        // may never proceed with set_irq_type() as the parent wait_for_interrupt() function will
+        // take the lock right after dropping one and hence set_irq_type() will wait for ever.
+        //
```

Comment:
What is the "parent" wait_for_interrupt() and why would it take the lock after dropping "one" - one what?  And why will set_irq_type() wait for ever when .edge_event_wait() uses a duration of one second?  This paragraph probably needs to be detailed a little more.

Reply:
Yes, it is not clear and still contain details of an earlier implentation that I had.

I am modifying it now as:

        // While waiting here for the interrupt to occur, it is possible that we receive another
        // request from the guest to disable the interrupt instead, via a call to set_irq_type()
        // (where we simply drop the request's buffer after acquiring the write lock for
        // `self.state[gpio]` first).
        //
        // Since this routine gets called indefinitely from GpioController::wait_for_interrupt(),
        // if we take the write lock (for `self.state[gpio]`) for the entire duration of this
        // routine, it is possible that the other thread (running set_irq_type()) may wait for ever
        // to take the lock.
        //
        // To avoid this situation, implement the state.request as an "Arc" instance instead and
        // take a reference to it without locking the state for the entire duration of the routine.
        // With this, while we are waiting for the interrupt, set_irq_type() can still proceed and
        // set the state.buffer to None and we will be able to quit the wait loop later on.